source-azure-blob-storage/main.go

```diff
@@ -0,0 +1,227 @@
+package main
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/estuary/connectors/filesource"
+	"github.com/estuary/flow/go/parser"
+	pf "github.com/estuary/flow/go/protocols/flow"
+)
+
+//go:embed schema.json
+var schema string
+
+type config struct {
+	AzureClientID       string         `json:"azureClientID"`
+	AzureClientSecret   string         `json:"azureClientSecret"`
+	AzureTenantID       string         `json:"azureTenantID"`
+	AzureSubscriptionID string         `json:"azureSubscriptionID"`
+	StorageAccountName  string         `json:"storageAccountName"`
+	ContainerName       string         `json:"containerName"`
+	Parser              *parser.Config `json:"parser"`
+}
+
+func (c config) Validate() error {
+	if c.AzureTenantID == "" {
+		return fmt.Errorf("missing AzureTenantID")
+	}
+	if c.AzureClientID == "" {
+		return fmt.Errorf("missing AzureClientID")
+	}
+	if c.AzureClientSecret == "" {
+		return fmt.Errorf("missing AzureClientSecret")
+	}
+	if c.AzureSubscriptionID == "" {
+		return fmt.Errorf("missing AzureSubscriptionID")
+	}
+	if c.StorageAccountName == "" {
+		return fmt.Errorf("missing StorageAccountName")
+	}
+	if c.ContainerName == "" {
+		return fmt.Errorf("missing ContainerName")
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.ContainerName
+}
+
+func (c config) RecommendedName() string {
+	return c.ContainerName
+}
+
+// TODO: Implement for Azure blob storage
+func (c config) FilesAreMonotonic() bool {
+	return true
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+// TODO: Implement for Azure blob storage
+func (c config) PathRegex() string {
+	return ""
+}
+
+type azureBlobStore struct {
+	client *azblob.Client
+	cfg    *config
+}
+
+// TODO: Test this
+func newAzureBlobStore(cfg config) (*azureBlobStore, error) {
+	os.Setenv("AZURE_CLIENT_ID", cfg.AzureClientID)
```

Comment:
Setting up authentication by setting env vars is pretty weird and I hope there is a cleaner way to do it. Seems like this is relying on the azure equivalent of AWS & friends' "application default credentials". But we don't use that for S3 and are instead able to do something like [this](https://github.com/estuary/connectors/blob/cd7c1b12ed2aa450e24953e2d5df18143f12e4ad/source-s3/main.go#L78-L107), and ideally we could configure the azure client in a similar way.